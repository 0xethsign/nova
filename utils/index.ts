/** Converts Ethereum wei to gwei. */
export const gweiToWei = (gwei: string | number) => {
  return 1e9 * parseFloat(gwei.toString());
};

/** Promise that resolves after `ms` millisecond.s */
export function sleep(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/** Calls `operation`, if it throws an error it will retry `retires` times waiting `delaySeconds` between each retry. */
export function retryOperation(operation: () => any, delaySeconds: number, retries: number) {
  return new Promise((resolve, reject) => {
    return operation()
      .then(resolve)
      .catch((reason) => {
        if (retries > 0) {
          return sleep(delaySeconds)
            .then(retryOperation.bind(null, operation, delaySeconds, retries - 1))
            .then(resolve)
            .catch(reject);
        }
        return reject(reason);
      });
  });
}
