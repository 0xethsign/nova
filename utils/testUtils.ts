import { ethers } from "hardhat";
import { ContractTransaction } from "ethers";

export const wait = async (tx: Promise<ContractTransaction>) => {
  const transaction = await tx;
  const receipt = await transaction.wait();

  return { transaction, receipt };
};

export const createTestWallet = (
  rpc: string,
  key = "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80"
) => {
  return new ethers.Wallet(key, new ethers.providers.JsonRpcProvider(rpc));
};

export function createFactory<T>(name: string): T {
  const artifact = require(`../artifacts/contracts/${name}.sol/${name}.json`);
  return new ethers.ContractFactory(artifact.abi, artifact.bytecode) as any;
}

export function createOVMFactory<T>(name: string): T {
  const artifact = require(`../artifacts-ovm/contracts/${name}.sol/${name}.json`);
  return new ethers.ContractFactory(artifact.abi, artifact.bytecode) as any;
}

export async function waitForL1ToL2Tx(
  tx: Promise<ContractTransaction>,
  watcher: any
) {
  console.log("Waiting for L1 -> L2 transaction...");
  const { transaction } = await wait(tx);
  const [msgHash] = await watcher.getMessageHashesFromL1Tx(transaction.hash);
  await watcher.getL2TransactionReceipt(msgHash);
  console.log("L1 -> L2 transaction completed!");
}
