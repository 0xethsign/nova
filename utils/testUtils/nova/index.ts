import { ethers } from "hardhat";
import { BigNumber } from "ethers";

/** Computes the execHash for a request like NovaExecHashLib. */
export function computeExecHash({
  nonce,
  strategy,
  calldata,
  gasPrice,
}: {
  nonce: number;
  strategy: string;
  calldata: string;
  gasPrice: number | BigNumber;
}) {
  return ethers.utils.solidityKeccak256(
    ["uint256", "address", "bytes", "uint256"],
    [nonce, strategy, calldata, gasPrice]
  );
}

export * from "./registry";
export * from "./executionManager";
