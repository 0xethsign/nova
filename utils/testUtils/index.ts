import chai from "chai";
import chaiAsPromised from "chai-as-promised";
import { jestSnapshotPlugin } from "mocha-chai-jest-snapshot";
chai.use(jestSnapshotPlugin());
chai.use(chaiAsPromised);
chai.should();

import { ethers } from "hardhat";
import { ContractReceipt, ContractTransaction } from "ethers";

import chalk from "chalk";
import { IERC20 } from "../../typechain";
import { Interface } from "ethers/lib/utils";

/** Gets an ethers factory for a contract. T should be the typechain factory type of the contract (ie: MockERC20__factory). */
export function getFactory<T>(name: string): Promise<T> {
  return ethers.getContractFactory(name) as any;
}

/** Increases EVM time by `seconds` and mines a new block. */
export async function increaseTimeAndMine(seconds: number) {
  await ethers.provider.send("evm_increaseTime", [seconds]);
  await ethers.provider.send("evm_mine", []);
}

/** Takes a contract interface and returns an array of all function sigHashes that are not pure/view.  */
export function getAllStatefulSigHashes(contractInterface: Interface) {
  return Object.entries(contractInterface.functions)
    .filter(([, fragment]) => {
      return !fragment.constant;
    })
    .map(([sig]) => contractInterface.getSighash(sig));
}

/** Records the gas usage of a transaction, and checks against the most recent saved Jest snapshot.
 * If not in CI mode it won't stop tests (just show a console log).
 * To update the Jest snapshot run `npm run gas-changed`
 */
export async function snapshotGasCost(x: Promise<ContractTransaction>) {
  let receipt: ContractReceipt = await (await x).wait();

  try {
    receipt.gasUsed.toNumber().should.toMatchSnapshot();
  } catch (e) {
    console.log(
      chalk.red(
        "(CHANGE) " +
          e.message
            .replace("expected", "used")
            .replace("to equal", "gas, but the snapshot expected it to use") +
          " gas"
      )
    );

    if (process.env.CI) {
      return Promise.reject("reverted: Gas consumption changed from expected.");
    }
  }

  return x;
}

/** Checkpoints `user`'s ether `token` balance upon calling.
 * Returns two functions (calcIncrease and calcDecrease,
 * calling calcIncrease will return the  `user`'s new `token`
 * balance minus the starting balance. Calling calcDecrease
 * subtracts the final balance from the balance.
 * */
export async function checkpointBalance(token: IERC20, user: string) {
  const startingBalance = await token.balanceOf(user);

  async function calcIncrease() {
    const finalBalance = await token.balanceOf(user);

    return finalBalance.sub(startingBalance);
  }

  async function calcDecrease() {
    const finalBalance = await token.balanceOf(user);

    return startingBalance.sub(finalBalance);
  }

  return [calcIncrease, calcDecrease];
}

export * from "./nova";
