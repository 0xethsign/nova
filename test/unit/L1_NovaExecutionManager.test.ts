import { createFactory, snapshotGasCost } from "../../utils/testUtils";

import { ethers } from "hardhat";
import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";

import {
  L2NovaRegistry__factory,
  L1NovaExecutionManager,
  L1NovaExecutionManager__factory,
  MockStrategy,
  MockCrossDomainMessenger,
  MockStrategy__factory,
  MockCrossDomainMessenger__factory,
  MockERC20,
  DSGuard,
  DSGuard__factory,
  MockERC20__factory,
} from "../../typechain";

describe("L1_NovaExecutionManager", function () {
  let signers: SignerWithAddress[];
  before(async () => {
    signers = await ethers.getSigners();
  });

  let L1_NovaExecutionManager: L1NovaExecutionManager;
  let DSGuard: DSGuard;

  /// Mocks
  let MockERC20: MockERC20;
  let MockStrategy: MockStrategy;
  let MockCrossDomainMessenger: MockCrossDomainMessenger;

  describe("constructor/setup", function () {
    it("should properly deploy mocks", async function () {
      const [deployer] = signers;

      MockERC20 = await createFactory<MockERC20__factory>(
        false,
        "MockERC20",
        "mocks/"
      )
        .connect(deployer)
        .deploy();

      MockStrategy = await createFactory<MockStrategy__factory>(
        false,
        "MockStrategy",
        "mocks/"
      )
        .connect(deployer)
        .deploy();

      MockCrossDomainMessenger =
        await createFactory<MockCrossDomainMessenger__factory>(
          false,
          "MockCrossDomainMessenger",
          "mocks/"
        )
          .connect(deployer)
          .deploy();
    });

    it("should properly deploy the execution manager", async function () {
      const [deployer] = signers;

      L1_NovaExecutionManager =
        await createFactory<L1NovaExecutionManager__factory>(
          false,
          "L1_NovaExecutionManager"
        )
          .connect(deployer)
          .deploy(
            ethers.constants.AddressZero,
            MockCrossDomainMessenger.address
          );
    });

    it("should properly use constructor arguments", async function () {
      // Make sure the constructor params were properly entered.
      await L1_NovaExecutionManager.messenger().should.eventually.equal(
        MockCrossDomainMessenger.address
      );
      await L1_NovaExecutionManager.L2_NovaRegistryAddress().should.eventually.equal(
        ethers.constants.AddressZero
      );
    });

    it("should contain constants that match expected values", async function () {
      // Make sure the hard revert text is correct.
      await L1_NovaExecutionManager.HARD_REVERT_TEXT().should.eventually.equal(
        "__NOVA__HARD__REVERT__"
      );

      // Make sure execCompletedMessageBytesLength is correct.
      await L1_NovaExecutionManager.execCompletedMessageBytesLength().should.eventually.equal(
        ((await createFactory<L2NovaRegistry__factory>(
          false,
          "L2_NovaRegistry"
        ).interface.encodeFunctionData("execCompleted", [
          ethers.utils.keccak256("0x00"),
          ethers.constants.AddressZero,
          0,
          false,
        ]).length) -
          2) /
          2
      );
    });

    describe("dsGuard", function () {
      it("should properly deploy a DSGuard", async function () {
        const [deployer] = signers;

        DSGuard = await createFactory<DSGuard__factory>(
          false,
          "DSGuard",
          "external/"
        )
          .connect(deployer)
          .deploy();
      });

      it("should properly init the owner", async function () {
        const [deployer] = signers;

        await DSGuard.owner().should.eventually.equal(deployer.address);
      });

      it("should allow setting the authorization of all functions to ANY", async function () {
        await DSGuard.permitBytes(
          await DSGuard.ANY(),
          await DSGuard.ANY(),
          await DSGuard.ANY()
        ).should.not.be.reverted;

        await DSGuard.canCall(
          ethers.constants.AddressZero,
          ethers.constants.AddressZero,
          L1_NovaExecutionManager.interface.getSighash(
            L1_NovaExecutionManager.interface.functions[
              "exec(uint72,address,bytes)"
            ]
          )
        ).should.eventually.equal(true);
      });

      it("should allow setting the owner to null", async function () {
        await DSGuard.setOwner(ethers.constants.AddressZero).should.not.be
          .reverted;

        await DSGuard.owner().should.eventually.equal(
          ethers.constants.AddressZero
        );
      });
    });

    describe("dsAuth", function () {
      it("should properly init the owner", async function () {
        const [deployer] = signers;

        await L1_NovaExecutionManager.owner().should.eventually.equal(
          deployer.address
        );
      });

      it("should allow connecting to a DSGuard", async function () {
        await L1_NovaExecutionManager.authority().should.eventually.equal(
          ethers.constants.AddressZero
        );

        await L1_NovaExecutionManager.setAuthority(DSGuard.address).should.not
          .be.reverted;
      });

      it("should allow setting the owner to null", async function () {
        await L1_NovaExecutionManager.setOwner(ethers.constants.AddressZero)
          .should.not.be.reverted;

        await L1_NovaExecutionManager.owner().should.eventually.equal(
          ethers.constants.AddressZero
        );
      });
    });
  });

  describe("hardRevert", function () {
    it("should revert with the proper message", async function () {
      await L1_NovaExecutionManager.hardRevert().should.be.revertedWith(
        await L1_NovaExecutionManager.HARD_REVERT_TEXT()
      );
    });
  });

  describe("exec/execWithRecipient", function () {
    it("should properly complete the first exec", async function () {
      await L1_NovaExecutionManager.exec(
        0,
        MockStrategy.address,
        MockStrategy.interface.encodeFunctionData("thisFunctionWillNotRevert")
      ).should.not.be.reverted;
    });

    it("hard revert should cause exec to revert", async function () {
      await L1_NovaExecutionManager.exec(
        1,
        MockStrategy.address,
        MockStrategy.interface.encodeFunctionData("thisFunctionWillHardRevert")
      ).should.be.revertedWith("HARD_REVERT");
    });

    it("soft revert should not exec to revert", async function () {
      await snapshotGasCost(
        L1_NovaExecutionManager.exec(
          2,
          MockStrategy.address,
          MockStrategy.interface.encodeFunctionData("thisFunctionWillRevert")
        )
      ).should.not.be.reverted;
    });

    it("should properly execute a minimal exec", async function () {
      await snapshotGasCost(
        L1_NovaExecutionManager.exec(
          2,
          MockStrategy.address,
          MockStrategy.interface.encodeFunctionData("thisFunctionWillNotRevert")
        )
      ).should.not.be.reverted;
    });

    it("should properly execute a stateful exec", async function () {
      await snapshotGasCost(
        L1_NovaExecutionManager.exec(
          3,
          MockStrategy.address,
          MockStrategy.interface.encodeFunctionData(
            "thisFunctionWillModifyState"
          )
        )
      ).should.not.be.reverted;

      await MockStrategy.counter().should.eventually.equal(2);
    });
  });

  describe("transferFromRelayer", function () {
    it("should transfer an arbitrary token to a strategy when requested", async function () {
      const [deployer] = signers;

      const weiAmount = ethers.utils.parseEther("1337");

      await MockERC20.approve(L1_NovaExecutionManager.address, weiAmount);

      await snapshotGasCost(
        L1_NovaExecutionManager.exec(
          4,
          MockStrategy.address,
          MockStrategy.interface.encodeFunctionData(
            "thisFunctionWillTransferFromRelayer",
            [MockERC20.address, weiAmount]
          )
        )
      )
        .should.emit(MockERC20, "Transfer")
        .withArgs(deployer.address, MockStrategy.address, weiAmount);

      await MockERC20.balanceOf(MockStrategy.address).should.eventually.equal(
        weiAmount
      );
    });
  });
});
