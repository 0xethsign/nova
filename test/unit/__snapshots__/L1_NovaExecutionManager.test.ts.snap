// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`L1_NovaExecutionManager exec/execWithRecipient should not revert due to a soft revert 1`] = `205163`;

exports[`L1_NovaExecutionManager exec/execWithRecipient should properly execute a minimal exec 1`] = `244180`;

exports[`L1_NovaExecutionManager exec/execWithRecipient should properly execute a stateful exec 1`] = `209372`;

exports[`L1_NovaExecutionManager transferFromRelayer should transfer an arbitrary token to a strategy when requested 1`] = `236330`;

exports[`L1_NovaExecutionManager transferFromRelayer will not allow a random contract to call during execution 1`] = `213907`;

exports[`L1_NovaExecutionManager transferFromRelayer will properly handle a transferFrom with no return value 1`] = `213782`;
